### PR TITLE
[test gap] Verify vlan interface autostate is disabled

### LINKED ISSUE #3601: Need new test: "no autostate" on VLAN interface
This is a test gap:

Need to confirm that vlan interface autostate is disabled.

### PR DESCRIPTION
**Pull request by lizhijianrd**
### Description of PR

Verify vlan interface autostate is **disabled** on SONiC.

The autostate feature notifies a switch or routing module VLAN interface (Layer 3 interface) to transition to up/up status when at least one Layer 2 port becomes active in that VLAN.

In SONiC, all vlans are bound to a single bridge interface, so the vlan interface will go down only if the bridge is down. Since bridge goes down when all the associated interfaces are down, if all the vlan members across all the vlans go down, the bridge will go down and the vlan interface will go down.

For more information about autostate, see: [Understanding and Troubleshooting the Autostate Feature in Catalyst Switches - Cisco](https://www.cisco.com/c/en/us/support/docs/switches/catalyst-6500-series-switches/41141-188.html)

Summary:
Fixes #3601

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012

#### What is the motivation for this PR?

Verify vlan interface autostate is **disabled** on SONiC.

#### How did you do it?

1. Find out all vlans which meet the following requirements:
    1. The `oper_state` of vlan interface is up;
    2. The `oper_state` of at least one member in the vlan is up.
2. Pick one vlan for test, and shutdown all the members in this vlan.
3. If more than one vlan was found in step 1, then we expect the vlan interface is still up. If only one vlan was found in step 1, then we expect the vlan interface is down.
4. Finally, restore all interfaces to their orignal `admin_state` (whether the test was passed, skipped, or failed).

#### How did you verify/test it?

Run this test case on vSONiC and physical DUTs.

### CODE REVIEW

File: tests/vlan/test_autostate_disabled.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        #   2. The oper_state of at least one member in the vlan is 'up'
+        vlan_available = []
+        for vlan in vlan_members_facts:
+            if ip_ifs.get(vlan, {}).get('oper_state') == 'up':
+                for member in vlan_members_facts[vlan]:
+                    if ifs_status.get(member, {}).get('oper') == 'up':
+                        vlan_available.append(vlan)
+                        break
+        if len(vlan_available) == 0:
+            pytest.skip('No vlan available on DUT {hostname}'.format(hostname=dut_hostname))
```
Comment (Blueve) [inline tests/vlan/test_autostate_disabled.py:53]: This message is ambiguous. Let's say:
`Not applicable VLAN available on DUT {hostname} for this test case`

File: tests/vlan/test_autostate_disabled.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+        # Pick a vlan for test
+        vlan = vlan_available[0]
+        vlan_members = vlan_members_facts[vlan].keys()
+
+        # Shutdown all the members in vlan.
+        res = duthost.shutdown_multiple(vlan_members)
+        if not res.is_successful:
+            self.restore_interface_admin_state(duthost, ifs_status)
+            pytest.fail('shutdown "{vlan_members}" in {vlan} failed'.format(vlan_members=vlan_members, vlan=vlan))
```
Comment (Blueve) [inline tests/vlan/test_autostate_disabled.py:63]: You can use this pattern, then you don't have to write the recover code everywhere.
```python
try:
   # test code
finally:
   self.restore...()
```

And please make sure intf's oper states are recovered before leave this test

File: tests/vlan/test_autostate_disabled.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        if not wait_until(60, 5, 0, self.check_interface_oper_state, duthost, vlan_members, "down"):
+            self.restore_interface_admin_state(duthost, ifs_status)
+            pytest.fail('shutdown "{vlan_members}" in {vlan} failed'.format(vlan_members=vlan_members, vlan=vlan))
+
+        # Check whether the oper_state of vlan interface is changed as expected.
+        ip_ifs = duthost.show_ip_interface()['ansible_facts']['ip_interfaces']
+        if len(vlan_available) > 1:
+            # If more than one vlan comply with the above test requirements, then there are members in other vlans
+            # that are still up. Therefore, the bridge is still up, and vlan interface should be up.
+            if ip_ifs.get(vlan, {}).get('oper_state') != "up":
```
Comment (Blueve) [inline tests/vlan/test_autostate_disabled.py:74]: Let's use `pytest_assert` directly

File: tests/vlan/test_autostate_disabled.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            if not res.is_successful:
+                logging.error('startup "{interfaces}" on DUT {hostname} failed'.
+                              format(interfaces=ifs_up, hostname=duthost.hostname))
+        if len(ifs_down) > 0:
+            res = duthost.shutdown_multiple(ifs_down)
+            if not res.is_successful:
+                logging.error('shutdown "{interfaces}" on DUT {hostname} failed'.
+                              format(interfaces=ifs_down, hostname=duthost.hostname))
+
+    def get_interface_status(self, duthost):
```
Comment (Blueve) [inline tests/vlan/test_autostate_disabled.py:109]: How about move this code the sonic host?

File: tests/vlan/test_autostate_disabled.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
+import logging
+import pytest
+
+from tests.common.utilities import wait_until
+
+
+pytestmark = [
+    pytest.mark.topology("t0")
```
Comment (lizhijianrd) [inline tests/vlan/test_autostate_disabled.py:8]: @Blueve Do you think t0 is a correct topology mark for this test case?

**Review by Blueve — CHANGES_REQUESTED** [review]

File: tests/vlan/test_autostate_disabled.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        if len(ifs_up) > 0:
+            res = duthost.no_shutdown_multiple(ifs_up)
+            if not res.is_successful:
+                logging.error('startup "{interfaces}" on DUT {hostname} failed'.
+                              format(interfaces=ifs_up, hostname=duthost.hostname))
+        if len(ifs_down) > 0:
+            res = duthost.shutdown_multiple(ifs_down)
+            if not res.is_successful:
+                logging.error('shutdown "{interfaces}" on DUT {hostname} failed'.
+                              format(interfaces=ifs_down, hostname=duthost.hostname))
```
Comment (lizhijianrd) [inline tests/vlan/test_autostate_disabled.py:107]: @Blueve In order not to affect subsequent test cases, I wonder if I have to wait planty of time to make sure the status of interfaces are fully restored.

File: tests/vlan/test_autostate_disabled.py (hunk 5 above)
Comment (Blueve) [reply tests/vlan/test_autostate_disabled.py:8]: Yes

File: tests/vlan/test_autostate_disabled.py (hunk 6 above)
Comment (Blueve) [reply tests/vlan/test_autostate_disabled.py:107]: let's use wati_until to make sure all intf states are expected

File: tests/vlan/test_autostate_disabled.py (hunk 5 above)
Comment (lizhijianrd) [reply tests/vlan/test_autostate_disabled.py:8]: OK, thanks!

File: tests/vlan/test_autostate_disabled.py (hunk 1 above)
Comment (lizhijianrd) [reply tests/vlan/test_autostate_disabled.py]: Thanks, I have fixed this in latest commit.

File: tests/vlan/test_autostate_disabled.py (hunk 2 above)
Comment (lizhijianrd) [reply tests/vlan/test_autostate_disabled.py]: Thanks, I have fixed this in latest commit.

File: tests/vlan/test_autostate_disabled.py (hunk 3 above)
Comment (lizhijianrd) [reply tests/vlan/test_autostate_disabled.py]: Thanks, I have fixed this in latest commit.

File: tests/vlan/test_autostate_disabled.py (hunk 4 above)
Comment (lizhijianrd) [reply tests/vlan/test_autostate_disabled.py]: Yes, I have moved this code and wrote an related doc here: https://github.com/Azure/sonic-mgmt/blob/4d4db8e6844677a46c79070aff5670151b295b61/docs/api_wiki/sonichost_methods/get_interfaces_status.md

File: tests/vlan/test_autostate_disabled.py (hunk 6 above)
Comment (lizhijianrd) [reply tests/vlan/test_autostate_disabled.py]: Thanks, I have added `wait_until` here in the latest commit.

File: docs/api_wiki/sonichost_methods/get_interfaces_status.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
+# get_interfaces_status
```
Comment (lizhijianrd) [inline docs/api_wiki/sonichost_methods/get_interfaces_status.md:1]: @Blueve I think this doc also need your review, thanks!

Comment (Blueve) [reply docs/api_wiki/sonichost_methods/get_interfaces_status.md:1]: lgtm

**Review by Blueve — APPROVED** [review]

**Pull request merged**